remotes/docker/fetcher.go

```diff
@@ -95,41 +95,51 @@ func (r dockerFetcher) Fetch(ctx context.Context, desc ocispec.Descriptor) (io.R
 			images.MediaTypeDockerSchema1Manifest,
 			ocispec.MediaTypeImageManifest, ocispec.MediaTypeImageIndex:
 
+			var lastErr error
 			for _, host := range r.hosts {
 				req := r.request(host, http.MethodGet, "manifests", desc.Digest.String())
 
 				rc, err := r.open(ctx, req, desc.MediaType, offset)
 				if err != nil {
-					if errdefs.IsNotFound(err) {
-						continue // try another host
+					// Store the error for referencing later
+					if lastErr == nil {
+						lastErr = err
 					}
-
-					return nil, err
+					continue // try another host
 				}
 
 				return rc, nil
 			}
+
+			if lastErr != nil {
+				return nil, lastErr
+			}
```

Comment:
Do you need to compare error here? You reaching here should mean "no-host could serve the content", shouldn't? So you can return the error immediately.
One case you get "lastErr == nil" here is "there is no host in r.host slice". I think we should detect this kind of failure in advance(before you enter this `for` loop)

And I think `lastError` isn't the correct name because the error stored in won't be the `last` one will be? :)

Reply:
Yes, hosts are checked prior to this code being run:
```go
	hosts := r.filterHosts(HostCapabilityPull)
	if len(hosts) == 0 {
		return nil, errors.Wrap(errdefs.ErrNotFound, "no pull hosts")
	}
```
Code updated to return error immediately. I've also changed `lastErr` to `firstErr` - hopefully this makes things a little clearer.

Reply:
Great! Thanks a lot.